lib/msf/core/post/file.rb

```diff
@@ -638,6 +642,153 @@ def copy_file(src_file, dst_file)
   end
   alias cp_file copy_file
 
+  private
+
+  def grab_icacls_groups(input_string)
+    final_groups_array = []
+    icacls_groups_array = input_string.scan(/^.+? (.+?\\.+?):/)
+    for i in 0...icacls_groups_array.length
+      final_groups_array << icacls_groups_array[i][0].strip # Strip any leading or trailing characters off of each of the groups found via the regex.
+    end
+    final_groups_array
+  end
+
+  #
+  # Parse a string, +input_string+, containing icacls output and filters it to find
+  # what groups are allowed to access the file that icacls was run on, and which
+  # permissions these groups have on that file.
+  #
+  # The resulting array that is returned is a multidimensional array where each
+  # element in the array is another array containing the name of the group as
+  # its first element, followed by an array containing the permissions
+  # that the group has on the file.
+  #
+  # @param input_string [String] String containing icacls output to parse for group permissions.
+  # @return [Array] Multidimensonal array containing the groups which have access to the file and what permissions they have on the file.
+  def grab_icacls_groups_permissions(input_string)
+    final_permissions_array = []
+    icacls_groups_permissions_array = input_string.scan(/\w:(\(.*\))$/)
+    for i in 0...icacls_groups_permissions_array.length
+      icacls_groups_permissions_array[i][0].strip! # Strip any leading or trailing characters off of each of the group permissions found via the regex.
+
+      # Permissions can exist inside a () section as comma delimited entries. To solve this we first remove the ( and ) characters,
+      # then we split the list on the delimiter, aka the comma character, before finally appending the resulting array to the end of
+      # final_permissions_array.
+      final_permissions_array << icacls_groups_permissions_array[i][0].gsub(/[()]/, '').split(',')
+    end
+    final_permissions_array
+  end
```

Comment:
```suggestion
  def find_icacls_permissions(path)
    icacls_output_raw = cmd_exec("icacls \"#{path}\"").to_s
    permissions_hash = {}
    icacls_output_raw.lines(chomp: true).each do |line|
      next unless line =~ /:\(/
      group, perms = line.gsub(path, '').gsub(/^\s+/, '').split(':')
      permissions_hash[group] = perms.gsub('(', '').gsub(',', ')').split(')')
    end
    permissions_hash
  end
```

Reply:
Suggested change implemented in https://github.com/rapid7/metasploit-framework/pull/15987/commits/21be7a0ee8a0de1cc3ab01a8d117819122a89120